utils/compilation_unit.mli

```diff
@@ -114,12 +114,18 @@ type argument =
     parameter name. *)
 val create_instance : t -> argument list -> t
 
-(** Create the compilation unit named by the given [Global_module.Name.t]. Only
-    meaningful if the global name is a _complete instantiation_, which is to say
-    either (a) the named module has no parameters, or (b) there is an argument
-    for each parameter and each argument is itself a complete instantiation.
-    This ensures the name determines a compile-time constant, and then the [t]
-    returned here is the module corresponding to that constant. *)
+(** Create the compilation unit named by the given [Global_module.Name.t].
+    Usually only meaningful if the global name is a _complete instantiation_,
+    which is to say either (a) the named module has no parameters, or (b) there
+    is an argument for each parameter and each argument is itself a complete
+    instantiation. This ensures the name determines a compile-time constant, and
+    then the [t] returned here is the module corresponding to that constant.
+
+    The exception to the above is that a global name with _no_ arguments makes
+    sense even if the named module takes parameters, only in this case the [t]
+    refers to the module with the instantiating functor. This is only relevant
+    in two cases: when we're compiling the parameterised module itself, and
+    when we're instantiating it. *)
 val of_global_name : Global_module.Name.t -> t
```

Comment:
`Name.t` is user-syntax; this directly interacting with `CU` is probably not good.

Can you leave a CR in `compilation_unit.mli`:

> Should make a type `Global_module.Complete.t` which represents a complete instantiation. It's isomorphic to `Global_module.Name.t`. Have a function: `val is_complete: Global_module.t -> Global_module.Complete.t option`. Then, CU should only interact with this new type.

Reply:
Makes sense to avoid mingling with user syntax, but for that we can just use `Global_module.t` instead, especially once the other caller can be covered by `Parameter.t` (once that's a thing). We can simply assert that the `Global_module.t` is complete. I'm pretty certain `Global_module.Complete.t` would only ever be used as an input to this function, since the whole point of a complete global is that it corresponds to a `CU.t`.

Reply:
Adding a CR saying more or less this (it should take a [Global_module.t] and check that it's complete, and there should also be `of_parameter`)